src/python/pants/engine/engine.py

```diff
@@ -111,8 +114,16 @@ def product_request(self, product, subjects):
     # TODO: See https://github.com/pantsbuild/pants/issues/3912
     throw_roots = tuple(root for root, state in result_items if type(state) is Throw)
     if throw_roots:
-      cumulative_trace = '\n'.join(self._scheduler.trace())
-      raise ExecutionError('Received unexpected Throw state(s):\n{}'.format(cumulative_trace))
+      # TODO test these cases
```

Comment:
Need to figure out the appropriate tests here. Also, this implementation feels ugly. I could back out the trace eliding, but that doesn't feel quite right either.